src/tools/pub/cmd/pub-add.md

```diff
@@ -26,65 +26,149 @@ For example, if 0.13.3 is the latest stable version of the `http` package,
-  For example, `dart pub add test http --dev` 
-  will add both the `test` and `http` packages 
-  as dev dependencies.
+{{site.alert.version-note}}
+  YAML-formatted descriptor syntax was added in Dart 2.19.
+  The descriptor replaces arguments like `--path`, `--sdk`, `--git-<option>`, etc.
+  We still support these arguments, but the documented method is now YAML-descriptor only.
 {{site.alert.end}}
 
-### `-d, --dev`
+The YAML descriptor reflects how dependencies are written in `pubspec.yaml`.
+
+```nocode
+$ dart pub add [options] [dev:]<package>[:descriptor] [[dev:]<package>[:descriptor]
+       ...]
+```
+
+Within the `descriptor` you can add a package with specific constraints or other sources:
+```nocode
+'<package>:{"<source>":"<descriptor>"[,"<source>":"<descriptor>"],"version":"<constraint>"}'
+```
+
+The `descriptor` syntax cannot be used in conjunction with any of the optional arguments it replaces.
+Their new corresponding YAML sources are listed below.
 
-Adds the package as a dev dependency,
+### `dev`
+
+Adds the package as a [dev dependency][],
 instead of as a regular dependency.
 
-### `--git-url=`_`<git_repo_url>`_
+[dev dependency]: /tools/pub/dependencies#dev-dependencies
+
+```terminal
+$ dart pub add dev:foo           # adds newest compatible stable version of foo
+$ dart pub add dev:foo:^2.0.0    # adds specified constraint of foo
+$ dart pub add foo dev:bar`      # adds regular dependency foo and dev dependency bar simultaneously
+```
+
+_Previously the `-d, --dev` option_:
+
+```terminal
+$ dart pub add --dev foo
+```
+
+### `git` 
+
+Adds a git dependency. 
+
+```terminal
+# dart pub add 'foo:{"git":"https://github.com/foo/foo"}'
+```
+
+You can also specify the repository, and the branch or commit, or exact location, within that repository:
+
+```terminal
+# dart pub add 'foo:{"git":{"url":"../foo.git","ref":"branch","path":"subdir"}}'
+```
+
+#### `url`
 
 Depends on the package in the
 [specified Git repository](/tools/pub/dependencies#git-packages).
 
+_Previously the `--git-url=<git_repo_url>` option_:
+
 ```terminal
 $ dart pub add http --git-url=https://github.com/my/http.git
 ```
 
-### `--git-ref=`_`<branch_or_commit>`_
+#### `ref`
 
-With `--git-url`, depends on the specified branch or commit of a Git repo.
+With `url`, depends on the specified branch or commit of a Git repo.
+
+_Previously the `--git-ref=<branch_or_commit>` option_:
 
 ```terminal
 $ dart pub add http --git-url=https://github.com/my/http.git --git-ref=tmpfixes
 ```
 
-### `--git-path=`_`<directory_path>`_
+#### `path`
+
+With `url`, specifies the location of a package within a Git repo.
+
+_Previously the `--git-path=<directory_path>` option_.
+
+### `hosted`
+
+Adds a hosted dependency that depends on
+the package server at the specified URL
+
+```terminal
+$ dart pub add 'foo:{"hosted":"my-pub.dev"}'
+```
+
+_Previously the `--hosted-url=<package_server_url>` option_.
+
+### `path`
+
+Adds path dependency on a locally stored package.
```

Comment:
Consider cross-linking path dependency or something else to [Path packages](https://dart.dev/tools/pub/dependencies#path-packages). Or something like "To learn more about path dependencies , see ...". There is a lot of useful context there on why you would use one as well as limitations.